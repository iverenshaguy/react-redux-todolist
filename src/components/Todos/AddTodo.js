import React, { Component } from 'react';
import PropTypes from 'prop-types';

class AddTodo extends Component {
  constructor() {
    super();
    this.state = {
      todo: {},
      value: ''
    };
  }

  handleChange = event => {
    this.setState({
      value: event.target.value
    });
  };

  handleSubmit = event => {
    event.preventDefault();
    if (this.state.value === '') {
      alert('Title is required');
      return;
    }

    if (this.props.addTodo) {
      this.setState(
        {
          todo: {
            userId: 1,
            id:
              this.props.todos.length >= 1
                ? +this.props.todos[this.props.todos.length - 1].id + 1
                : 1,
            title: this.state.value,
            completed: false
          }
        },
        () => {
          this.props.addTodo(this.state.todo);
        }
      );
    } else if (this.props.editTodo) {
      const editedTodo = {
        userId: 1,
        id: this.props.todo.id,
        title: this.state.value,
        completed: false
      };
      this.setState(
        prevState => ({
          todo: Object.assign({}, this.state.todo, editedTodo)
        }),
        () => {
          this.props.editTodo(this.state.todo);
        }
      );
    }
  };

  render() {
    return (
      <div className="AddTodo">
        <form className="form-inline my-2 my-lg-0" onSubmit={this.handleSubmit}>
          <input
            className="form-control mr-sm-2 new-todo"
            type="text"
            name="todoinput"
            value={this.state.value}
            onChange={this.handleChange}
            placeholder="Enter Todo"
            autoFocus={true}
          />
          <button type="submit" className="btn btn-primary my-2 my-sm-0">
            {this.props.buttonText}
          </button>
        </form>
      </div>
    );
  }
}

AddTodo.proptypes = {
  todo: PropTypes.object,
  addTodo: PropTypes.func,
  editTodo: PropTypes.func,
  buttonText: PropTypes.string.isRequired
};

export default AddTodo;
