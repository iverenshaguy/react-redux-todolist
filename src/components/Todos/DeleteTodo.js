import React from 'react';
import PropTypes from 'prop-types';
import Button from '../shared/Button';

const DeleteTodo = props => {
  return (
    <Button
      buttonType="button"
      btnClass="close"
      buttonText="&times;"
      onClick={props.onClick}
    />
  );
};

DeleteTodo.proptypes = {
  onClick: PropTypes.func.isRequired
};

export default DeleteTodo;
