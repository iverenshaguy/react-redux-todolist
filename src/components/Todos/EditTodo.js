import React from 'react';
import PropTypes from 'prop-types';
import Button from '../shared/Button';

const EditTodo = props => {
  return (
    <Button
      buttonType="button"
      btnClass="btn btn-outline-primary"
      buttonText="Edit Todo"
      onClick={props.onClick}
    />
  );
};

EditTodo.proptypes = {
  onClick: PropTypes.func.isRequired
};

export default EditTodo;
