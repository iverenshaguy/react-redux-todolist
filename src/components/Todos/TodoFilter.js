import React from 'react';
import FilterButton from '../../containers/FilterButton';

const SHOW_ALL = 'all';
const SHOW_COMPLETED = 'completed';
const SHOW_UNCOMPLETED = 'uncompleted';

const TodoFilter = () => {
  return (
    <div>
      <FilterButton filter={SHOW_ALL}>All</FilterButton>&nbsp;&nbsp;
      <FilterButton filter={SHOW_UNCOMPLETED}>
        Uncompleted
      </FilterButton>&nbsp;&nbsp;
      <FilterButton filter={SHOW_COMPLETED}>Completed</FilterButton>&nbsp;&nbsp;
    </div>
  );
};

export default TodoFilter;
