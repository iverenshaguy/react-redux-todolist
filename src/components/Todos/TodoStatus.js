import React from 'react';
import PropTypes from 'prop-types';
import Button from '../shared/Button';

const TodoStatus = props => {
  let buttonText;
  if (props.todo.completed) {
    buttonText = 'Mark as Undone';
  } else {
    buttonText = 'Mark as Done';
  }

  return (
    <Button
      buttonType="button"
      btnClass="btn btn-outline-primary"
      buttonText={buttonText}
      onClick={props.onClick}
    />
  );
};

TodoStatus.proptypes = {
  todo: PropTypes.object.isRequired,
  onEdit: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired
};

export default TodoStatus;
