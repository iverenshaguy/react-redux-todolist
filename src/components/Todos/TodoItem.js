import React, { Component } from 'react';
import PropTypes from 'prop-types';
import AddTodo from './AddTodo';
import EditTodo from './EditTodo';
import DeleteTodo from './DeleteTodo';
import TodoStatus from './TodoStatus';
import './TodoItem.css';

class TodoItem extends Component {
  constructor() {
    super();
    this.state = {
      todoTitle: 'show',
      editForm: 'hide'
    };
  }

  editTodoToggle = () => {
    if (this.state.editForm === 'hide') {
      this.setState({
        todoTitle: 'hide',
        editForm: 'show'
      });
    } else {
      this.setState({
        todoTitle: 'show',
        editForm: 'hide'
      });
    }
  };

  handleEditTodo = todo => {
    const updateTodo = new Promise((resolve, reject) => {
      this.props.onEdit(todo);
      resolve();
    });

    updateTodo.then(() => {
      this.setState({
        todoTitle: 'show',
        editForm: 'hide'
      });
    });
  };

  handleTodoStatus = () => {
    const { todo } = this.props;
    const newTodo = {
      userId: 1,
      id: todo.id,
      title: todo.title,
      completed: !todo.completed
    };

    this.props.onEdit(newTodo);
  };

  handleDeleteTodo = () => {
    const id = this.props.todo.id;
    this.props.onDelete(id);
  };

  render() {
    const { editForm, todoTitle } = this.state;
    const { todo } = this.props;

    let completed;

    if (todo.completed) {
      completed = 'Completed';
    } else {
      completed = 'Not Completed';
    }

    return (
      <div className="TodoItem">
        <div className="card border-primary">
          <div className="card-body">
            <DeleteTodo onClick={this.handleDeleteTodo} />
            <blockquote className="card-blockquote">
              <p className={todoTitle}>
                {todo.title}&nbsp;&nbsp;
                <span className="badge badge-primary">{completed}</span>
              </p>
              <div className={editForm}>
                <AddTodo
                  todo={todo}
                  buttonText="Update Todo"
                  editTodo={this.handleEditTodo}
                />
              </div>
              <br />
              <footer>
                <EditTodo onClick={this.editTodoToggle} />&nbsp;&nbsp;
                <TodoStatus todo={todo} onClick={this.handleTodoStatus} />
              </footer>
            </blockquote>
          </div>
        </div>
        <br />
      </div>
    );
  }
}

TodoItem.proptypes = {
  todo: PropTypes.object.isRequired,
  onEdit: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired
};

export default TodoItem;
