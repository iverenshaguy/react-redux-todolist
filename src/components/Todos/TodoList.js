import React from 'react';
import PropTypes from 'prop-types';
import TodoItem from './TodoItem';

const TodoList = props => {
  const {
    todos,
    activeTodo,
    editTodoToggle,
    handleDeleteTodo,
    handleTodoStatus
  } = props;

  const todoItems = todos
    .slice(0)
    .reverse()
    .map(todo => {
      return (
        <div key={todo.id} id={`todoitem-${todo.id}`}>
          <TodoItem
            todo={todo}
            activeTodo={activeTodo}
            editTodoToggle={() => editTodoToggle(todo.id)}
            handleDeleteTodo={() => handleDeleteTodo(todo.id)}
            handleTodoStatus={() => handleTodoStatus(todo.id)}
          />
        </div>
      );
    });

  return (
    <div>
      <h3>Todos ({todos.length})</h3>
      <hr />
      {todoItems}
    </div>
  );
};

TodoList.defaultProps = {
  activeTodo: null
};

TodoList.proptypes = {
  todos: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      completed: PropTypes.bool.isRequired,
      title: PropTypes.string.isRequired
    }).isRequired
  ).isRequired,
  activeTodo: PropTypes.number,
  editTodoToggle: PropTypes.func.isRequired,
  handleDeleteTodo: PropTypes.func.isRequired,
  handleTodoStatus: PropTypes.func.isRequired
};

export default TodoList;
