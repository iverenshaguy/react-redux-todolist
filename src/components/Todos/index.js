import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import AddTodo from './AddTodo';
import TodoItem from './TodoItem';
import Button from '../shared/Button';

const SHOW_ALL = 'SHOW_ALL';
const SHOW_COMPLETED = 'SHOW_COMPLETED';
const SHOW_UNCOMPLETED = 'SHOW_UNCOMPLETED';

class Todos extends Component {
  constructor() {
    super();
    this.state = {
      todos: [],
      sort: SHOW_ALL
    };
  }

  handleAddTodo = newTodo => {
    this.props.api.createTodo(newTodo).then(todo =>
      this.setState(prevState => ({
        todos: [...prevState.todos, todo],
        sort: SHOW_ALL
      }))
    );
  };

  handleEditTodo = todo => {
    this.props.api.updateTodo(todo).then(updatedTodo => {
      const newTodos = this.state.todos.map(
        todoItem =>
          updatedTodo.id === todoItem.id
            ? Object.assign({}, todoItem, updatedTodo)
            : todoItem
      );

      this.setState({
        todos: newTodos,
        todoTitle: 'show',
        editForm: 'hide'
      });
    });
  };

  handleDeleteTodo = id => {
    const newTodos = this.state.todos.filter(todo => todo.id !== id);

    this.props.api.deleteTodo(id).then(() => {
      this.setState({
        todos: newTodos,
        sort: SHOW_ALL
      });
    });
  };

  showAllTodos = () => {
    this.setState({
      sort: SHOW_ALL
    });
  };

  showUncompletedTodos = () => {
    this.setState({
      sort: SHOW_UNCOMPLETED
    });
  };

  showCompletedTodos = () => {
    this.setState({
      sort: SHOW_COMPLETED
    });
  };

  componentDidMount = () => {
    this.props.api.getTodos().then(todos =>
      this.setState({
        todos: todos
      })
    );
  };

  renderNoTodo = () => {
    return <p>You don't have any todos yet. Add a Todo</p>;
  };

  renderTodos = () => {
    const { todos, sort } = this.state;
    let sortedTodos;

    if (sort === SHOW_ALL) {
      sortedTodos = todos;
    }

    if (sort === SHOW_UNCOMPLETED) {
      sortedTodos = todos.filter(todo => !todo.completed);
    }

    if (sort === SHOW_COMPLETED) {
      sortedTodos = todos.filter(todo => todo.completed);
    }

    const todoItems = sortedTodos
      .slice(0)
      .reverse()
      .map(todo => {
        return (
          <div key={todo.id} id={`todoitem-${todo.id}`}>
            <TodoItem
              todo={todo}
              onEdit={this.handleEditTodo}
              onDelete={this.handleDeleteTodo}
            />
          </div>
        );
      });

    return todoItems;
  };

  render() {
    const { todos, sort } = this.state;
    const showAllBtnClass = classNames({
      btn: true,
      'show-all-btn': true,
      'btn-outline-primary': sort !== SHOW_ALL,
      'btn-primary': sort === SHOW_ALL
    });
    const showCompletedBtnClass = classNames({
      btn: true,
      'show-completed-btn': true,
      'btn-outline-primary': sort !== SHOW_COMPLETED,
      'btn-primary': sort === SHOW_COMPLETED
    });
    const showUncompletedBtnClass = classNames({
      btn: true,
      'show-uncompleted-btn': true,
      'btn-outline-primary': sort !== SHOW_UNCOMPLETED,
      'btn-primary': sort === SHOW_UNCOMPLETED
    });

    return (
      <div className="Todos">
        <div className="header">
          <h1>Todos</h1>
          <br />
          <h3>Add Todo Item</h3>
          <AddTodo
            todos={todos}
            addTodo={this.handleAddTodo}
            buttonText="Add Todo"
          />
          <hr />
          <h3>Todos ({todos.length})</h3>
          <hr />
          <Button
            btnClass={showAllBtnClass}
            buttonText="All"
            onClick={this.showAllTodos}
          />&nbsp;&nbsp;
          <Button
            btnClass={showUncompletedBtnClass}
            buttonText="Uncompleted"
            onClick={this.showUncompletedTodos}
          />&nbsp;&nbsp;
          <Button
            btnClass={showCompletedBtnClass}
            buttonText="Completed"
            onClick={this.showCompletedTodos}
          />&nbsp;&nbsp;
        </div>
        <hr />
        <div className="todolist">
          {todos.length === 0 ? this.renderNoTodo() : this.renderTodos()}
        </div>
        <br />
      </div>
    );
  }
}

Todos.proptypes = {
  api: PropTypes.shape({
    getTodos: PropTypes.func.isRequired,
    createTodo: PropTypes.func.isRequired,
    updateTodo: PropTypes.func.isRequired,
    deleteTodo: PropTypes.func.isRequired
  })
};

export default Todos;
