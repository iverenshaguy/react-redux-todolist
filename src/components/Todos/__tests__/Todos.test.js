import Todos from '../index';
import { todos as mockTodos } from '../../../__mocks__/todos.mock';
import todosApi from '../../../helpers/api/__mocks__/todos.api';

// jest.mock('../../../helpers/api/__mocks__/todos.api');

describe('Todos', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Todos api={todosApi} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('mounts correctly', () => {
    const wrapper = mount(<Todos api={todosApi} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('correctly updates the state after AJAX call in `componentDidMount` was made', done => {
    const wrapper = mount(<Todos api={todosApi} />);
    expect.assertions(1);
    setImmediate(() => {
      expect(wrapper.instance().state.todos.length).toBe(4);
      done();
    });
  });

  it('shows all todos', () => {
    const wrapper = mount(<Todos api={todosApi} />);
    wrapper.setState({
      todos: mockTodos,
      sort: 'SHOW_COMPLETED'
    });

    wrapper.find('.show-all-btn').simulate('click');
    expect(wrapper.instance().state.sort).toEqual('SHOW_ALL');
    expect(wrapper.find('.todolist').text()).toContain(' Levies and I ');
    expect(wrapper.find('.todolist').text()).toContain(' and on time ');
    expect(wrapper.find('.todolist').text()).toContain(' fly less come ');
    expect(wrapper.find('.todolist').text()).toContain(
      ' Who will be easy for those services '
    );
  });

  it('shows only completed todos', () => {
    const wrapper = mount(<Todos api={todosApi} />);
    wrapper.setState({
      todos: mockTodos,
      sort: 'SHOW_ALL'
    });

    wrapper.find('.show-completed-btn').simulate('click');
    expect(wrapper.instance().state.sort).toEqual('SHOW_COMPLETED');
    expect(wrapper.find('.todolist').text()).toContain(' Levies and I ');
    expect(wrapper.find('.todolist').text()).toContain(' and on time ');
    expect(wrapper.find('.todolist').text()).not.toContain(' fly less come ');
    expect(wrapper.find('.todolist').text()).not.toContain(
      ' Who will be easy for those services '
    );
  });

  it('shows only uncompleted todos', () => {
    const wrapper = mount(<Todos api={todosApi} />);
    wrapper.setState({
      todos: mockTodos,
      sort: 'SHOW_ALL'
    });

    wrapper.find('.show-uncompleted-btn').simulate('click');
    expect(wrapper.instance().state.sort).toEqual('SHOW_UNCOMPLETED');
    expect(wrapper.find('.todolist').text()).toContain(' fly less come ');
    expect(wrapper.find('.todolist').text()).toContain(
      ' Who will be easy for those services '
    );
    expect(wrapper.find('.todolist').text()).not.toContain(' Levies and I ');
    expect(wrapper.find('.todolist').text()).not.toContain(' and on time ');
  });

  it('toggles todo', () => {
    const wrapper = mount(<Todos api={todosApi} />);
    const showAllTodosSpy = jest.spyOn(wrapper.instance(), 'showAllTodos');
    const showCompletedTodosSpy = jest.spyOn(
      wrapper.instance(),
      'showCompletedTodos'
    );
    const showUncompletedTodosSpy = jest.spyOn(
      wrapper.instance(),
      'showUncompletedTodos'
    );

    wrapper.setState({
      todos: mockTodos,
      sort: 'SHOW_ALL'
    });

    wrapper.find('.show-all-btn').simulate('click');
    expect(showAllTodosSpy).toHaveBeenCalled();

    wrapper.find('.show-completed-btn').simulate('click');
    expect(showCompletedTodosSpy).toHaveBeenCalled();

    wrapper.find('.show-uncompleted-btn').simulate('click');
    expect(showUncompletedTodosSpy).toHaveBeenCalled();
  });

  it('adds a new todo', done => {
    const wrapper = mount(<Todos api={todosApi} />);
    const event = { target: { value: 'Learn React TDD' } };

    setImmediate(() => {
      expect(wrapper.instance().state.todos.length).toBe(4);
      wrapper.find('.header AddTodo input').simulate('change', event);
      wrapper.find('.header AddTodo form').simulate('submit');

      setImmediate(() => {
        expect(wrapper.instance().state.todos.length).toBe(5);
        expect(wrapper.text()).toContain('Learn React TDD');
        done();
      });
    });
  });

  it('edits a todo', done => {
    const wrapper = mount(<Todos api={todosApi} />);
    const event = { target: { value: 'Learn React TDD with Jest and Enzyme' } };

    setImmediate(() => {
      wrapper.update();
      wrapper.find('#todoitem-2 input').simulate('change', event);
      wrapper.find('#todoitem-2 form').simulate('submit');
      setImmediate(() => {
        expect(wrapper.instance().state.todos[1].title).toBe(
          'Learn React TDD with Jest and Enzyme'
        );
        expect(wrapper.text()).toContain(
          'Learn React TDD with Jest and Enzyme'
        );
        done();
      });
    });
  });

  it('changes Todo status', done => {
    const wrapper = mount(<Todos api={todosApi} />);
    setImmediate(() => {
      wrapper.update();
      wrapper.find('#todoitem-2 TodoStatus').simulate('click');
      setImmediate(() => {
        expect(wrapper.instance().state.todos[1].completed).toBeTruthy();
        done();
      });
    });
  });

  it('deletes a todo', done => {
    const wrapper = mount(<Todos api={todosApi} />);
    setImmediate(() => {
      wrapper.update();
      wrapper.find('#todoitem-2 DeleteTodo').simulate('click');
      setImmediate(() => {
        expect(wrapper.instance().state.todos.length).toBe(3);
        done();
      });
    });
  });
});
