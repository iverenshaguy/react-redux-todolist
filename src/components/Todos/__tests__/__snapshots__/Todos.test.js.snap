// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos mounts correctly 1`] = `
<Todos
  api={
    Object {
      "createTodo": [Function],
      "deleteTodo": [Function],
      "getTodos": [Function],
      "updateTodo": [Function],
    }
  }
>
  <div
    className="Todos"
  >
    <div
      className="header"
    >
      <h1>
        Todos
      </h1>
      <br />
      <h3>
        Add Todo Item
      </h3>
      <AddTodo
        addTodo={[Function]}
        buttonText="Add Todo"
        todos={Array []}
      >
        <div
          className="AddTodo"
        >
          <form
            className="form-inline my-2 my-lg-0"
            onSubmit={[Function]}
          >
            <input
              autoFocus={true}
              className="form-control mr-sm-2 new-todo"
              name="todoinput"
              onChange={[Function]}
              placeholder="Enter Todo"
              type="text"
              value=""
            />
            <button
              className="btn btn-primary my-2 my-sm-0"
              type="submit"
            >
              Add Todo
            </button>
          </form>
        </div>
      </AddTodo>
      <hr />
      <h3>
        Todos (
        0
        )
      </h3>
      <hr />
      <Button
        btnClass="btn show-all-btn btn-primary"
        buttonText="All"
        onClick={[Function]}
      >
        <button
          className="btn show-all-btn btn-primary"
          onClick={[Function]}
          type="button"
        >
          All
        </button>
      </Button>
        
      <Button
        btnClass="btn show-uncompleted-btn btn-outline-primary"
        buttonText="Uncompleted"
        onClick={[Function]}
      >
        <button
          className="btn show-uncompleted-btn btn-outline-primary"
          onClick={[Function]}
          type="button"
        >
          Uncompleted
        </button>
      </Button>
        
      <Button
        btnClass="btn show-completed-btn btn-outline-primary"
        buttonText="Completed"
        onClick={[Function]}
      >
        <button
          className="btn show-completed-btn btn-outline-primary"
          onClick={[Function]}
          type="button"
        >
          Completed
        </button>
      </Button>
        
    </div>
    <hr />
    <div
      className="todolist"
    >
      <p>
        You don't have any todos yet. Add a Todo
      </p>
    </div>
    <br />
  </div>
</Todos>
`;

exports[`Todos renders correctly 1`] = `
<div
  className="Todos"
>
  <div
    className="header"
  >
    <h1>
      Todos
    </h1>
    <br />
    <h3>
      Add Todo Item
    </h3>
    <AddTodo
      addTodo={[Function]}
      buttonText="Add Todo"
      todos={Array []}
    />
    <hr />
    <h3>
      Todos (
      0
      )
    </h3>
    <hr />
    <Button
      btnClass="btn show-all-btn btn-primary"
      buttonText="All"
      onClick={[Function]}
    />
      
    <Button
      btnClass="btn show-uncompleted-btn btn-outline-primary"
      buttonText="Uncompleted"
      onClick={[Function]}
    />
      
    <Button
      btnClass="btn show-completed-btn btn-outline-primary"
      buttonText="Completed"
      onClick={[Function]}
    />
      
  </div>
  <hr />
  <div
    className="todolist"
  >
    <p>
      You don't have any todos yet. Add a Todo
    </p>
  </div>
  <br />
</div>
`;
